KCFTOOLS
---
KCFTOOLS is a Java-based tool designed to identify genomic variations by analyzing k-mer counts between reference and query sequences. The tool leverages k-mer counting databases, such as those generated by KMC, to perform a variety of analyses, including variation detection, IBS window identification, and genotype matrix generation.

---

## Features
- **Screen for Variations**: Detect variations by comparing reference k-mers against a KMC database.
- **Cohort Creation**: Create a cohort of sample KCF files for downstream analysis.
- **IBS Window Identification**: Detect Identity-by-State (IBS) windows in a KCF file.
- **Chromosome-wise Splitting**: Split KCF files based on chromosomes for targeted analyses.
- **Attribute Extraction**: Extract detailed attributes from KCF files.
- **Genotype Matrix Generation**: Convert KCF files into genotype matrices for use in population studies.

---

## Requirements
- Java 11 or higher
- KMC (K-mer Counting tool) for generating k-mer databases

---

## Installation
1. Clone the repository:
   ```bash
   git clone <repository-url>
   cd kcftools
2. Build the project:
   ```bash
   mvn clean package
3. The compiled JAR file will be located in the `target` directory.
4. Run the JAR file:
   ```bash
   java -jar target/kcftools-<version>.jar
   
---


## Usage
# General Usage
```bash
java -jar kcftools-<version>.jar <command> [options]
```
# Commands
The list of available commands and their descriptions are as follows:
1. **`getVariations`**: Detect variations by comparing reference k-mers against a KMC database.
* Usage:
```bash
java -jar kcftools-<version>.jar getVariations [options]
  -k, --kmc=<kmcDBprefix>    KMC database prefix
  -m, --memory               Load KMC database into memory
  -o, --output=<outFile>     Output file name
  -r, --reference=<refFasta> Reference file name
  -s, --sample=<sampleName>  Sample name
  -t, --threads=<nThreads>   Number of threads
  -w, --window=<windowSize>  Window size
```
2. **`cohort`**: Create a cohort of sample KCF files for downstream analysis.
* Usage:
```bash
java -jar kcftools-<version>.jar cohort [options]
  -i, --input=<inFiles>[,<inFiles>...]
                           List of samples kcf files
  -l, --list=<listFile>    File containing list of samples kcf files
  -o, --output=<outFile>   Output file name
 ```
3. **`findIBS`**: Detect Identity-by-State (IBS) windows in a KCF file.
* Usage:
```bash
java -jar kcftools-<version>.jar findIBS [options]
      --bed                Write bed file
  -i, --input=<inFile>     Input KCF file name
      --min=<minConsecutive>
                           Minimum number of consecutive windows
  -o, --output=<outFile>   Output KCF file name
  -r, --reference=<refFasta>
                           Reference Fasta file name
      --score=<scoreCutOff>
                           Score cut-off
      --summary            Write summary tsv file
      --var                Detect Variable Regions instead of IBS
```
4. **`splitKCF`**: Split KCF files based on chromosomes for targeted analyses.
* Usage:
```bash
java -jar kcftools-<version>.jar splitKCF [options]
  -k, --kcf=<kcfFile>     KCF file name
  -o, --output=<outDir>   Output directory
```
5. **`getAttributes`**: Extract detailed attributes from KCF files.
* Usage:
```bash
java -jar kcftools-<version>.jar getAttributes [options]
  -i, --input=<kcfFile>    KCF file name
  -o, --output=<outFile>   Output file name prefix
```
6. **`genotypeMatrix`**: Convert KCF files into genotype matrices for use in population studies.
* Usage:
```bash
java -jar kcftools-<version>.jar genotypeMatrix [options]
  -i, --input=<inFile>       Input KCF file
  -o, --output=<outPrefix>   Output file prefix
```

> WARNING: Tests are still under development. Please use the tool with caution.

## Contact
For any queries, please contact siva.selvanayagam@wur.nl
